src/library/scala/util/hashing/MurmurHash3.scala

```diff
@@ -96,31 +96,49 @@ private[hashing] class MurmurHash3 {
    *  This is useful for hashing sets, for example.
    */
   final def unorderedHash(xs: TraversableOnce[Any], seed: Int): Int = {
-    var a, b, n = 0
-    var c = 1
-    xs foreach { x =>
-      val h = x.##
-      a += h
-      b ^= h
-      if (h != 0) c *= h
-      n += 1
+    if  (xs.isEmpty) {
+      var h = seed
+      h = mix(h, 0)
+      h = mix(h, 0)
+      h = mixLast(h, 1)
+      finalizeHash(h, 0)
+    }
+    else {
+      object hasher extends Function1[Any, Unit] {
+        var a, b, n = 0
+        var c = 1
+        override def apply(x: Any): Unit = {
+          val h = x.##
+          a += h
+          b ^= h
+          if (h != 0) c *= h
+          n += 1
+        }
+      }
+      xs foreach hasher
+      var h = seed
+      h = mix(h, hasher.a)
+      h = mix(h, hasher.b)
+      h = mixLast(h, hasher.c)
+      finalizeHash(h, hasher.n)
```

Comment:
An alternative to consider is:

```
val iterator = xs.iterator()
while (iterator.hasNext() {
  ...
}
```

That would also just be a single allocation of the iterator. (Or no allocation when the collection is empty). That might be the simplest structure of this code when we merge this forward to the 2.13.x branch that already calls `iterator.`

I'm happy enough with this version of the code for 2.12.x though -- I'll make the change in the 2.13.x branch when I merge this forward.